File: frontend/src/interpreter/interpreter.ts
import { AST } from "./ast";
import { Lexer } from "./lexer";
import { Parser } from "./parser";

export class Interpreter {
  private parser: Parser;

  constructor(parser: Parser) {
    this.parser = parser;
  }

  public interpret(): any[] {
    let tree: AST = this.parser.parse();
    return tree.eval();
  }
}

/**
 * interpret source code and output result
 * @param text raw input text
 */
export function interpretSource(text: string): any[] | string {
  let lexer = new Lexer(text);
  let parser = new Parser(lexer);
  let interpreter = new Interpreter(parser);
  try {
    return interpreter.interpret();
  } catch (exception) {
    return exception.message;
  }
}

File: mathscript/interpreter.cpp
//
// Created by Pranav Putta on 9/22/20.
//

#include "lexer.h"
#include "parser.h"
#include "ast.h"
#include "interpreter.h"

std::string mathscript::interpret(std::string input) {
    Lexer lexer(input);
    Parser parser(lexer);
    ASTPtr root = parser.Parse();
    Result res = root->eval(SymTable::ConstructGlobalTable());

    std::string out;
    for (auto &c: std::get<std::vector<ObjPtr >>(res.data)) {
        auto num = dynamic_cast<Number *>(c.get());
        out += std::to_string(num->value) + "\n";
    }
    return out;
}

Napi::String mathscript::interpretWrapped(const Napi::CallbackInfo &info) {
    Napi::Env env = info.Env();
    //check if arguments are integer only.
    if (info.Length() < 2 || !info[0].IsString()) {
        Napi::TypeError::New(env, "arg1::String expected").ThrowAsJavaScriptException();
    }
    //convert javascripts datatype to c++
    Napi::String first = info[0].As<Napi::String>();
    //run c++ function return value and return it in javascript
    Napi::String returnValue = Napi::String::New(env, mathscript::interpret(first.Utf8Value()));

    return returnValue;
}

Napi::Object mathscript::Init(Napi::Env env, Napi::Object exports) {
    //export Napi function
    exports.Set("interpret", Napi::Function::New(env, mathscript::interpretWrapped));
    return exports;
}